packages/protocol/contracts/bridge/Bridge.sol

```diff
@@ -448,9 +448,16 @@ contract Bridge is EssentialContract, IBridge {
     /// @return The minimal gas limit required for sending this message.
     function getMessageMinGasLimit(uint256 dataLength) public pure returns (uint32) {
         unchecked {
-            // Message struct takes 7*32=224 bytes + a variable length array.
-            // Since ABI.encode pads data to multiples of 32 bytes, we over-charge 32 bytes
-            return GAS_RESERVE + uint32((dataLength + 256) >> 4);
+            // The abi encoding of A = (Message calldata msg) is 10 * 32 bytes
+            // + 32 bytes (A is a dynamic tuple, offset to first elements)
+            // + 32 bytes (offset to last bytes element of Message)
+            // + 32 bytes (padded encoding of length of Message.data + dataLength (padded to 32
+            // bytes)
+            // = 13 * 32 + (dataLength / 32 * 32) + 32.
+            // non-zero calldata cost per byte is 16.
+
+            uint256 dataCost = (dataLength / 32 * 32 + 448) << 4;
```

Comment:
Does using a bitshift actually save gas? I would think the compiler just handles it

Reply:
I think it does, but not verified. We also got such optimization suggestions from code4rena auditors.